sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/RowToPubsubMessage.java

```diff
@@ -57,14 +73,43 @@ public static RowToPubsubMessage withTimestampAttribute(boolean useTimestampAttr
             ? input.apply(WithTimestamps.of((row) -> row.getDateTime(TIMESTAMP_FIELD).toInstant()))
             : input;
 
-    return withTimestamp
-        .apply(DropFields.fields(TIMESTAMP_FIELD))
-        .apply(ToJson.of())
-        .apply(
-            MapElements.into(TypeDescriptor.of(PubsubMessage.class))
-                .via(
-                    (String json) ->
-                        new PubsubMessage(
-                            json.getBytes(StandardCharsets.ISO_8859_1), ImmutableMap.of())));
+    withTimestamp = withTimestamp.apply(DropFields.fields(TIMESTAMP_FIELD));
+    switch (payloadFormat) {
+      case JSON:
+        return withTimestamp
+            .apply("MapRowToJsonString", ToJson.of())
+            .apply("MapToJsonBytes", MapElements.via(new StringToBytes()))
+            .apply("MapToPubsubMessage", MapElements.via(new ToPubsubMessage()));
+      case AVRO:
+        return withTimestamp
+            .apply(
+                "MapRowToAvroBytes",
+                MapElements.via(AvroUtils.getRowToAvroBytesFunction(payloadSchema)))
+            .apply("MapToPubsubMessage", MapElements.via(new ToPubsubMessage()));
+      default:
+        throw new IllegalArgumentException("Unsupported payload format: " + payloadFormat);
+    }
+  }
+
+  private static class StringToBytes extends SimpleFunction<String, byte[]> {
+    @Override
+    public byte[] apply(String s) {
+      return s.getBytes(ISO_8859_1);
```

Comment:
I'm not sure why I opted to get ISO_8859_1 encoded bytes here... Could you change this to use UTF_8?

Reply:
Sure! Done.